File: DFS/129.sum-root-to-leaf-numbers.go
package leetcode

import "strconv"

// @leet start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sumNumbers(root *TreeNode) int {
	var ret int
	var dfs func(node *TreeNode, cur string)
	dfs = func(node *TreeNode, cur string) {
		if node == nil {
			// always need, consider node only had left, the right will hit nil
			// need skip right nil node.
			return
		}
		curString := strconv.Itoa(node.Val)
		if node.Left == nil && node.Right == nil {
			num, _ := strconv.Atoi(cur + curString)
			ret += num
			return
		}
		dfs(node.Left, cur+curString)
		dfs(node.Right, cur+curString)
	}
	dfs(root, "")
	return ret
}

// @leet end

File: DFS/129.sum-root-to-leaf-numbers.py
from typing import Optional
from type import TreeNode


# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        res = 0

        def dfs(n: TreeNode | None, s: str):
            if n is None:  # NOTE: required , consider node with only one child
                return
            nonlocal res
            if n.left is None and n.right is None:
                res += int(s + str(n.val))
                return
            s += str(n.val)
            dfs(n.left, s)
            dfs(n.right, s)

        dfs(root, "")
        return res


# @leet end

